diesel/src/sqlite/connection/stmt.rs

```diff
@@ -189,88 +188,113 @@ impl Drop for Statement {
+    // We use a dyn here just to erase the generic type
+    _query: Box<dyn QueryFragment<Sqlite> + 'query>,
 }
 
 impl<'stmt, 'query> BoundStatement<'stmt, 'query> {
-    fn bind<T>(
-        mut statement: MaybeCached<'stmt, Statement>,
-        query: T,
-    ) -> QueryResult<BoundStatement<'stmt, 'query>>
+    fn bind<T>(statement: MaybeCached<'stmt, Statement>, query: T) -> QueryResult<Self>
     where
         T: QueryFragment<Sqlite> + QueryId + 'query,
     {
-        // Don't use a trait object here to prevent using a virtual function call
-        // For sqlite this can introduce a measurable overhead
-        let mut query = ManuallyDrop::new(query);
-
-        let mut bind_collector = SqliteBindCollector::new();
-        query.collect_binds(&mut bind_collector, &mut ())?;
-
-        let SqliteBindCollector { metadata, binds } = bind_collector;
-        let mut binds_to_free = ManuallyDrop::new(Vec::new());
-
-        for (bind_idx, (bind, tpe)) in (1..).zip(binds.into_iter().zip(metadata)) {
-            // It's safe to call bind here as:
-            // * The type and value matches
-            // * We ensure that corresponding buffers lives long enough below
-            // * The statement is not used yet by `step` or anything else
-            let res = unsafe { statement.bind(tpe, &bind, bind_idx) };
-
-            if let Err(e) = res {
-                Self::unbind_buffers(&mut statement, &binds_to_free);
-
-                unsafe {
-                    // This is safe as we return early from this function, we
-                    // unbound the binds above, so it's fine now to drop
-                    // the owned binds
-                    ManuallyDrop::drop(&mut binds_to_free);
-                    // At last we can also drop query
-                    ManuallyDrop::drop(&mut query);
-                }
-                return Err(e);
-            }
+        // Query is boxed here to make sure it won't move in memory anymore, so any bind
+        // it could output would stay valid.
+        let query: Box<T> = Box::new(query);
+
+        let mut bind_collector = SqliteBindCollector::<'query>::new();
+        (*query).collect_binds(&mut bind_collector, &mut ())?;
+
+        // We create Self as early as possible to make sure Drop impl is called however
+        // we exit from this function (error or panic)
+        //
+        // After this point we will be moving the boxed query around, but that won't change the
+        // validity of the collected binds because they were built from the query stored on the
+        // heap, which will not be deallocated or moved without the Self we're building being
+        // `Drop`ped, erasing any corresponding reference
+        let bind_collector = unsafe {
+            std::mem::transmute::<SqliteBindCollector<'_>, SqliteBindCollector<'static>>(
+                bind_collector,
+            )
+        };
```

Comment:
While this is technically safe in the scope of the current function I really try to avoid `transmute` as much as possible, even if it makes the code a bit harder to write. My issue with `transmute` is that any code using `transmute` may be valid now, but it's incredible hard to spot if that remains true after a refactoring, especially for persons who've not written the original code.

Reply:
That is just to put the query in Self asap.
We're calling a function that makes use of these references as static references right after anyway, which may become as incorrect. 
I think it's reasonable to assume that changes to this file are hard enough to make that they would be made by people who have read the whole file and would be thoroughly reviewed.

Reply:
You are correct, doing that here is fine. I consider this to be a minor issue for now, the main point is the discussion above about aliasing.